cpp/src/Ice/ConnectionI.cpp

```diff
@@ -1476,25 +1430,35 @@ Ice::ConnectionI::message(ThreadPoolCurrent& current)
 
 // executeFromThisThread dispatches to the correct DispatchQueue
 #ifdef ICE_SWIFT
-    _threadPool->executeFromThisThread(make_shared<ExecuteUpcall>(
-        shared_from_this(),
-        std::move(connectionStartCompleted),
-        sentCBs,
-        std::move(messageUpcall),
-        current.stream));
+    auto stream = make_shared<InputStream>();
+    current.stream.swap(*stream);
+
+    auto self = shared_from_this();
+    _threadPool->executeFromThisThread(
+        [self,
+         connectionStartCompleted = std::move(connectionStartCompleted),
+         sentCBs = std::move(sentCBs),
+         messageUpcall = std::move(messageUpcall),
+         stream]() { self->upcall(connectionStartCompleted, sentCBs, std::move(messageUpcall), *stream); },
+        self);
 #else
     if (!_hasExecutor) // Optimization, call dispatch() directly if there's no executor.
     {
         upcall(connectionStartCompleted, sentCBs, std::move(messageUpcall), current.stream);
```

Comment:
I would also move connectionStartedCompleted in all these calls for consistency.